x/merkledb/codec.go

```diff
@@ -154,25 +156,23 @@ func (c *codecImpl) decodeDBNode(b []byte, n *dbNode, branchFactor BranchFactor)
 	switch {
 	case err != nil:
 		return err
-	case numChildren > uint64(branchFactor):
-		return errTooManyChildren
 	case numChildren > uint64(src.Len()/minChildLen):
 		return io.ErrUnexpectedEOF
 	}
 
-	n.children = make(map[byte]child, branchFactor)
+	n.children = make(map[byte]child, numChildren)
 	var previousChild uint64
 	for i := uint64(0); i < numChildren; i++ {
 		index, err := c.decodeUint(src)
 		if err != nil {
 			return err
 		}
-		if index >= uint64(branchFactor) || (i != 0 && index <= previousChild) {
-			return errChildIndexTooLarge
+		if i != 0 && index <= previousChild || index > math.MaxUint8 {
```

Comment:
semantically the same but makes operator precedence more clear
```suggestion
		if (i != 0 && index <= previousChild) || index > math.MaxUint8 {
```